intermediate_source/TP_tutorial.rst

```diff
@@ -0,0 +1,282 @@
+When the Transformer model becomes larger (i.e. Llama2 70B), it will also quickly hit the limitation 2: One could not use FSDP alone with even local batch_size=1 due to memory and convergence constraints. For example, Llama 2 global batch size is 1k, so data parallelism alone can not be used at 2k GPUs.
+
+
+How to apply Tensor Parallel
+---------------
+
+
+To demonstrate how to apply tensor parallelism using the PyTorch native Tensor Parallel APIs, let us look at a common Transformer model. In this tutorial, we use the most recent `Llama2 model <https://github.com/pytorch/examples/blob/main/distributed/tensor_parallelism/>__` as a reference Transformer model implementation, as it is also widely used in the community.
+
+Since Tensor Parallel shards individual tensors over a set of devices, we would need to set up the distributed environment (i.e NCCL communicators) first. Tensor Parallelism is a SPMD sharding algorithm similar to PyTorch DDP/FSDP, and it under the hood leverages the PyTorch DTensor to perform sharding. It also utilizes the DeviceMesh abstraction (which under the hood manages ProcessGroups) for device management and sharding. To see how to utilize DeviceMesh to set up multi-dimensional parallelisms, please refer to this tutorial. Tensor Parallel usually works within each host, so let us first initialize a DeviceMesh that connects 8 GPUs within a host.
+
+.. code-block:: python
+    # run this via torchrun: torchrun --standalone --nproc_per_node=8 ./tp_tutorial.py
+
+    from torch.distributed.device_mesh import init_device_mesh
+
+    tp_mesh = init_device_mesh("cuda", (8,))
+
+
+Now we initialize the DeviceMesh, let us take a detailed look at the Llama2 model architecture and see how we should perform the tensor parallel sharding. Here we focus on the core TransformerBlock, where the Transformer model stacks the identical TransformerBlocks to scale up the model.
+
+The core TransformerBlock consists of an Attention layer and a FeedForward layer. Let us first look at the simpler FeedForward layer. For the FeedForward Layer it consists of three Linear layers, where it performs a SwiGLU style MLP, looking at its forward function:
+
+.. code-block:: python
+
+    def forward(self, x):
+        return self.w2(F.silu(self.w1(x)) * self.w3(x))
+
+
+It performs ``w1`` and ``w3`` matmuls concurrently and followed by a w2 matmul with the result of the combined w1/w3 linear projection results. This means we could use the idea from the Tensor Parallelism paper to shard the w1/w3 Linear layers in the colwise fashion and shard the w2 Linear layer in the rowwise fashion, so that there iss only one allreduce communication happening at the end of all the three layers. With the PyTorch native Tensor Parallel, we can simply create a parallelize_plan for the FeedForward layer like below:
```

Comment:
```suggestion
It performs ``w1`` and ``w3`` matmuls concurrently and followed by a w2 matmul with the result of the combined w1/w3 linear projection results. This means we could use the idea from the Tensor Parallelism paper to shard the w1/w3 Linear layers in the colwise fashion and shard the w2 Linear layer in the rowwise fashion, so that there is only one ``allreduce`` communication happening at the end of all the three layers. With the PyTorch native Tensor Parallel, we can simply create a ``parallelize_plan`` for the Feed Forward layer like below:
```

Reply:
pls consider making naming consistent: e.g. w2 -> \`\`w2\`\`